python/lsst/daf/butler/core/repoTransfers.py

```diff
@@ -437,17 +437,36 @@ def load(self, datastore: Datastore, *,
         fileDatasets = []
         for (datasetTypeName, run), records in self.datasets.items():
             datasetType = self.registry.getDatasetType(datasetTypeName)
+            # Insert all DatasetRefs for all FileDatasets together, in one
+            # big flattened list.
+            # For now, we ignore the dataset_id we pulled from the file
+            # and just insert without one to get a new autoincrement value.
+            # Eventually (once we have origin in IDs) we'll preserve them.
+            refs = self.registry.insertDatasets(
+                datasetType,
+                # Iterator comprehension ordering with two 'for' clauses seems
+                # backward, but this does indeed loop over records first and
+                # then fileDataset.refs (which is of course the only order that
+                # makes any sense).
+                dataIds=(ref.dataId for fileDataset, _ in records for ref in fileDataset.refs),
+                run=run,
+                recursive=True
+            )
+            # Now iterate over the original records, and install the new
+            # resolved DatasetRefs to replace the unresolved ones as we
+            # reorganize the collection information.
+            i = 0
             for fileDataset, collectionsForDataset in records:
-                # For now, we ignore the dataset_id we pulled from the file
-                # and just insert without one to get a new autoincrement value.
-                # Eventually (once we have origin in IDs) we'll preserve them.
-                fileDataset.refs = [self.registry.addDataset(datasetType, ref.dataId, run=run,
-                                                             recursive=True) for ref in fileDataset.refs]
+                n = len(fileDataset.refs)
```

Comment:
This is a bit messy, maybe it would be better do build the "n"s in a list while you generate the input to insertDatasets, and then just loop over those accumulated "n"s here